lib/openai/client.rb

```diff
@@ -52,37 +52,62 @@ def translate(parameters: {})
     end
 
     def self.get(path:)
-      HTTParty.get(
-        uri(path: path),
-        headers: headers,
-        timeout: request_timeout
-      )
+      to_json(conn.get(uri(path: path), timeout: OpenAI.configuration.request_timeout) do |req|
+        req.headers = headers
+      end&.body)
     end
 
     def self.json_post(path:, parameters:)
-      HTTParty.post(
-        uri(path: path),
-        headers: headers,
-        body: parameters&.to_json,
-        timeout: request_timeout
-      )
+      to_json(conn.post(uri(path: path), timeout: OpenAI.configuration.request_timeout) do |req|
+        if parameters[:stream].is_a?(Proc)
+          req.options.on_data = to_json_stream(user_proc: parameters[:stream])
+          parameters[:stream] = true # Necessary to tell OpenAI to stream.
+        end
+
+        req.headers = headers
+        req.body = parameters.to_json
+      end&.body)
     end
 
     def self.multipart_post(path:, parameters: nil)
-      HTTParty.post(
-        uri(path: path),
-        headers: headers.merge({ "Content-Type" => "multipart/form-data" }),
-        body: parameters,
-        timeout: request_timeout
-      )
+      m_conn = Faraday.new { |f| f.request :multipart }
+
+      to_json(m_conn.post(uri(path: path), timeout: OpenAI.configuration.request_timeout) do |req|
+        req.headers = headers.merge({ "Content-Type" => "multipart/form-data" })
+        req.body = multipart_parameters(parameters)
+      end&.body)
     end
 
     def self.delete(path:)
-      HTTParty.delete(
-        uri(path: path),
-        headers: headers,
-        timeout: request_timeout
-      )
+      to_json(conn.delete(uri(path: path), timeout: OpenAI.configuration.request_timeout) do |req|
+        req.headers = headers
+      end&.body)
+    end
+
+    def self.to_json(string)
+      return unless string
+
+      JSON.parse(string)
+    rescue JSON::ParserError
+      # Convert a multiline string of JSON objects to a JSON array.
+      JSON.parse(string.gsub("}\n{", "},{").prepend("[").concat("]"))
+    end
+
+    private_class_method def self.to_json_stream(user_proc:)
+      proc do |chunk, bytesize|
+        # Clean up response string of chunks.
+        chunk = chunk.gsub("\n\ndata: [DONE]\n\n", "").gsub("\n\ndata:", ",").gsub("data: ", "")
+
+        # Turn it into JSON.
+        chunk = to_json(chunk)
```

Comment:
@alexrudall I tried this branch locally and it's working for the most part, but I'm occasionally seeing the following error thrown in this `to_json` call:
```ruby
451: unexpected token at 'DONE]\n\n]'
```

It seems like the `gsub` call above might be missing some cases, although I wasn't able to get to the root cause yet, I'll take another look and see if I can consistently reproduce this but just wanted to raise in case you've seen this error (also see you pushed some new commits so it might have been fixed already).

Reply:
Instead of the calls to `gsub` it seems like it would only be necessary to capture data if it looks like a JSON object. You could do someting like `json_data = chunk.match(/data: (\{.*\})/i)` and only attempt to parse the data if `json_data` is present. If it's not, it would be `[DONE]` or some other non-json like object. I think their API docs suggest it should either be a JSON object, i.e. `{...}` or `[DONE]` so those chunks that do not match could be ignored. This would simplify this code a bit and handle the situation found by @fwitzke.

Reply:
I'm working on a writing some specs and was able to reproduce 

```
JSON::ParserError:
       unexpected token at 'foo]'
```

Basically just pass "foo" as a chunk and it will try to parse it, fail, and then attempt to turn it into an array of JSON objects which is where the `]` comes from. I'll try to wrap this up today and submit for review.

Reply:
Also while trying to repro this I got one error response from openai API, it doesn't seem like the code handles that right now, this is the `chunk` data received:
```
data: {"id":"chatcmpl-79HdSUgAA1wTHnw8fKIboQcadubqK","object":"chat.completion.chunk","created":1682447098,"model":"gpt-3.5-turbo-0301","choices":[{"delta":{},"index":0,"finish_reason":"stop"}]}

data: {"error":{"message":"The server experienced an error while processing your request. Sorry about that! You can retry your request, or contact us through our help center at help.openai.com if the error persists.","type":"server_error","param":null,"code":null}}

data: [DONE]

```

Reply:
For what it's worth I did make an attempt here: https://github.com/alexrudall/ruby-openai/pull/248. Your feedback is very welcome and I'm happy to adjust if I'm on an agreeable track.

Reply:
that simplifies a lot @rmontgomery429, I think that change makes sense and would address both cases I mentioned!

Reply:
Thanks for the excellent work on this. Merged without comment!